pkg/issuer/acme/dns/dns.go

```diff
@@ -155,6 +156,25 @@ func (s *Solver) solverFor(crt *v1alpha1.Certificate, domain string) (solver, er
 		if err != nil {
 			return nil, fmt.Errorf("error instantiating route53 challenge solver: %s", err.Error())
 		}
+	case providerConfig.AzureDNS != nil:
+		clientSecret, err := s.secretLister.Secrets(s.resourceNamespace).Get(providerConfig.AzureDNS.ClientSecret.Name)
+		if err != nil {
+			return nil, fmt.Errorf("error getting azuredns client secret: %s", err.Error())
+		}
+
+		clientSecretBytes, ok := clientSecret.Data[providerConfig.AzureDNS.ClientSecret.Key]
+		if !ok {
+			return nil, fmt.Errorf("error getting azure dns client secret: key '%s' not found in secret", providerConfig.AzureDNS.ClientSecret.Key)
+		}
+
+		impl, err = azuredns.NewDNSProviderCredentials(
+			providerConfig.AzureDNS.ClientID,
+			string(clientSecretBytes),
+			providerConfig.AzureDNS.SubscriptionID,
+			providerConfig.AzureDNS.TenantID,
+			providerConfig.AzureDNS.ResourceGroupName,
+			providerConfig.AzureDNS.HostedZoneName,
```

Comment:
We should add a check somewhere early in the authz flow to ensure these variables are set. If we perform the check after requesting an authorization, we will enter a loop and use up a users quota.

I can't remember off the top of my head if we have a spot in the code where this is done already. We really need to make an explicit validations package.